File: app/src/main/java/com/example/dices2/KeyboardActivity.java
package com.example.dices2;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import static com.example.dices2.Consts.*;

public class KeyboardActivity extends AppCompatActivity {

    private TextView textViewTitle;
    private TextView textViewResult;
    private TextView buttonSave;
    private TextView buttonPlus;
    private String result = "";
    private boolean isEdit;
    private boolean isInSchool;
    private int enabledTextColor;
    private int disabledTextColor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_keyboard);
        init();
        Intent intent = getIntent();
        String rowName = intent.getStringExtra(INTENT_ROW_NAME);
        textViewTitle.setText(rowName);
        isEdit = intent.getBooleanExtra(INTENT_IS_EDIT, false);
        isInSchool = intent.getBooleanExtra(INTENT_IS_IN_SCHOOL, false);
        if (isEdit) {
            result = getIntent().getStringExtra(INTENT_EDITED_VALUE);
            textViewResult.setText(result);
        }
        enableButton(buttonSave, isEdit);
        enableButton(buttonPlus, isInSchool);
    }

    private void init() {
        textViewTitle = findViewById(R.id.tv_keyboard_title);
        textViewResult = findViewById(R.id.tv_keyboard_result);
        buttonSave = findViewById(R.id.b_save);
        buttonPlus = findViewById(R.id.b_plus);
        enabledTextColor = getResources().getColor(R.color.white);
        disabledTextColor = getResources().getColor(R.color.colorBackgroundLight);
    }

    public void onKeyboardButtonClicked(View view) {
        String symbol = ((TextView)view).getText().toString();
        boolean isResultEmpty = result.equals("");
        boolean isFirstResultSymbolPlusOrMinus = result.equals("+") || result.equals("-");
        boolean isSymbolPlusOrMinus = symbol.equals("+") || symbol.equals("-");
        boolean isSymbolZero = symbol.equals("0");
        if ((isResultEmpty && !isSymbolZero) ||
                (isFirstResultSymbolPlusOrMinus && !isSymbolPlusOrMinus && !isSymbolZero) ||
                (!isResultEmpty && !isFirstResultSymbolPlusOrMinus && !isSymbolPlusOrMinus)) {
            result += symbol;
            textViewResult.setText(result);
            enableButton(buttonSave, true);
        }
    }

    public void onBackspaceButtonClicked(View view) {
        if (!result.equals("")) {
            result = result.substring(0, result.length() - 1);
            textViewResult.setText(result);
        }
        if (result.equals("") && !isEdit) {
            enableButton(buttonSave, false);
        }
    }

    public void onCancelButtonClicked(View view) {
        finish();
    }

    public void onSaveButtonClicked(View view) {
        Intent intent = new Intent();
        intent.putExtra(Consts.INTENT_INPUT_VALUE, result);
        intent.putExtra(INTENT_IS_EDIT, isEdit);
        setResult(RESULT_OK, intent);
        finish();
    }

    private void enableButton(TextView textView, boolean isEnabled) {
        textView.setTextColor(isEnabled ? enabledTextColor : disabledTextColor);
        textView.setEnabled(isEnabled);
    }
}

File: app/src/main/java/com/example/dices2/KeyboardActivity.kt
package com.example.dices2

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.example.dices2.databinding.ActivityKeyboardBinding

class KeyboardActivity : AppCompatActivity() {

    private lateinit var binding: ActivityKeyboardBinding

    private var result = ""
    private var isEdit = false
    private var isInSchool = false
    private var enabledTextColor = 0
    private var disabledTextColor = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityKeyboardBinding.inflate(layoutInflater)
        setContentView(binding.root)
        init()

        val intent = intent
        val rowName = intent.getStringExtra(Consts.INTENT_ROW_NAME)
        binding.tvKeyboardTitle.text = rowName
        isEdit = intent.getBooleanExtra(Consts.INTENT_IS_EDIT, false)
        isInSchool = intent.getBooleanExtra(Consts.INTENT_IS_IN_SCHOOL, false)
        if (isEdit) {
            getIntent().getStringExtra(Consts.INTENT_EDITED_VALUE)?.let { result = it }
            binding.tvKeyboardResult.text = result
        }
        enableButton(textView = binding.bSave, isEnabled = isEdit)
        enableButton(textView = binding.bPlus, isEnabled = isInSchool)
        if (isInSchool) {
            enableDigitButtons(isEnabled = false)
        } else {
            enableDigitButtons(isEnabled = true)
        }
    }

    private fun init() {
        // TODO getColor deprecated
        enabledTextColor = resources.getColor(R.color.white)
        disabledTextColor = resources.getColor(R.color.colorBackgroundLight)
    }

    fun onKeyboardButtonClicked(view: View) {
        val symbol = (view as TextView).text.toString()
        val isCountKeyboardInputSymbolsLessThanThree = result.length < 3
        val isResultEmpty = result == ""
        val isFirstResultSymbolPlusOrMinus = result == "+" || result == "-"
        val isSymbolPlusOrMinus = symbol == "+" || symbol == "-"
        val isSymbolZero = symbol == "0"
        // TODO disable to input school without first '+' or '-'
        val isFirstSymbolInSchoolPlusOrMinus = (isInSchool && isResultEmpty && isSymbolPlusOrMinus) || isFirstResultSymbolPlusOrMinus
        if ((isResultEmpty && !isSymbolZero ||
                        isFirstResultSymbolPlusOrMinus && !isSymbolPlusOrMinus && !isSymbolZero ||
                        !isResultEmpty && !isFirstResultSymbolPlusOrMinus && !isSymbolPlusOrMinus)
                && isCountKeyboardInputSymbolsLessThanThree) {
            result += symbol
            binding.tvKeyboardResult.text = result
            enableButton(textView = binding.bSave, isEnabled = true)
            enableDigitButtons(isEnabled = true)
        }
    }

    fun onBackspaceButtonClicked(view: View?) {
        if (result != "") {
            result = result.substring(0, result.length - 1)
            binding.tvKeyboardResult.text = result
        }
        if (result == "" && !isEdit) {
            enableButton(textView = binding.bSave, isEnabled = false)
            if (isInSchool) {
                enableDigitButtons(isEnabled = false)
            }
        }
    }

    fun onCancelButtonClicked(view: View?) {
        finish()
    }

    fun onSaveButtonClicked(view: View?) {
        val intent = Intent()
        intent.putExtra(Consts.INTENT_INPUT_VALUE, result)
        intent.putExtra(Consts.INTENT_IS_EDIT, isEdit)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    private fun enableButton(textView: TextView, isEnabled: Boolean) {
        textView.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        textView.isEnabled = isEnabled
    }

    private fun enableDigitButtons(isEnabled: Boolean) {
        binding.b0.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b1.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b2.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b3.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b4.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b5.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b6.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b7.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b8.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b9.setTextColor(if (isEnabled) enabledTextColor else disabledTextColor)
        binding.b0.isEnabled = isEnabled
        binding.b1.isEnabled = isEnabled
        binding.b2.isEnabled = isEnabled
        binding.b3.isEnabled = isEnabled
        binding.b4.isEnabled = isEnabled
        binding.b5.isEnabled = isEnabled
        binding.b6.isEnabled = isEnabled
        binding.b7.isEnabled = isEnabled
        binding.b8.isEnabled = isEnabled
        binding.b9.isEnabled = isEnabled
    }
}